### PR TITLE
optimize scalar sampling

### PR DESCRIPTION
**Pull request by daming-lu**

### CODE REVIEW

File: server/visualdl/lib.py [hunk 1]
```diff
@@ -30,6 +30,10 @@ def get_scalar_tags(storage, mode):
 
 
 def get_scalar(storage, mode, tag, num_records=300):
+    # if sample amount is negative, 0 or 1, use default
+    if num_records <= 1:
```
Comment (Superjomn) [inline server/visualdl/lib.py]: assert num_records > 1

serious error.

File: server/visualdl/lib.py [hunk 2]
```diff
@@ -41,13 +45,23 @@ def get_scalar(storage, mode, tag, num_records=300):
...
 
-        span = float(len(data) / num_records)
-        end_idx = len(data) - 1
-        res = []
-        for i in xrange(num_records):
-            id = int(end_idx - i * span)
-            res.append(data[id])
-        return [v for v in reversed(res)]
+        # if sample amount is larger than actual data amount, then choose all data
+        if num_records > len(data):
```
Comment (Superjomn) [inline server/visualdl/lib.py]: conflicts with line 45

no need to sample again.

File: server/visualdl/lib.py [hunk 3]
```diff
@@ -41,13 +45,23 @@ def get_scalar(storage, mode, tag, num_records=300):
...
+            num_records = len(data)
+
+        span = float(len(data)) / (num_records - 1)
+        span_multiply = 0
+
+        data_idx = int(span_multiply * span)
```
Comment (Superjomn) [inline server/visualdl/lib.py]: span_multiply -> span_offset ?

File: server/visualdl/lib.py [hunk 4]
```diff
@@ -41,13 +45,23 @@ def get_scalar(storage, mode, tag, num_records=300):
...
+        sampled_data = []
+
+        while data_idx < len(data):
+            sampled_data.append(data[data_idx])
+            span_multiply += 1
+            data_idx = int(span_multiply * span)
+
+        sampled_data.append(data[-1])
```
Comment (Superjomn) [inline server/visualdl/lib.py]: sample from data[-1] to data[0], so the latest record can be taken naturally.

File: server/visualdl/lib.py (hunk 1 above)
Comment (daming-lu) [reply server/visualdl/lib.py]: done

File: server/visualdl/lib.py (hunk 2 above)
Comment (daming-lu) [reply server/visualdl/lib.py]: done

File: server/visualdl/lib.py (hunk 3 above)
Comment (daming-lu) [reply server/visualdl/lib.py]: done

File: server/visualdl/lib.py (hunk 4 above)
Comment (daming-lu) [reply server/visualdl/lib.py]: done

File: server/visualdl/lib.py [hunk 5]
```diff
@@ -41,13 +43,19 @@ def get_scalar(storage, mode, tag, num_records=300):
...
-            res.append(data[id])
-        return [v for v in reversed(res)]
+        span = float(len(data)) / (num_records - 1)
+        span_offset = 0
+
+        data_idx = int(span_offset * span)
+        sampled_data = []
+
+        while data_idx < len(data):
+            sampled_data.append(data[len(data) - data_idx - 1])
```
Comment (Superjomn) [inline server/visualdl/lib.py]: len(...) will run every time in the iter
place it outside the whileloop like:

```
data_size = len(data)

while ...:
    ...
```

Comment (daming-lu) [reply server/visualdl/lib.py]: done

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**